pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go

```diff
@@ -50,38 +50,112 @@ func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.Envo
+			return fmt.Errorf("function MergeAnyWithAny failed for ApplyClusterMerge: %v", errMerge)
+		}
+
+		// Merge the above result with the whole cluster
+		merge.Merge(dstCluster, retVal)
+	}
+	return nil
+}
+
+// Applies special case for patching Transport Socket and Transport Socket Matches
+// and removes used TS and TSM structs from source patch
+func mergeTransportSocketCluster(c *cluster.Cluster, cp *model.EnvoyFilterConfigPatchWrapper) (err error) {
 	cpValueCast, okCpCast := (cp.Value).(*cluster.Cluster)
 	if !okCpCast {
-		return false, fmt.Errorf("cast of cp.Value failed: %v", okCpCast)
+		return fmt.Errorf("cast of cp.Value failed: %v", okCpCast)
+	}
+
+	// First merge transport socket matches with the same name
+	if (len(c.GetTransportSocketMatches())) > 0 && (len(cpValueCast.GetTransportSocketMatches()) > 0) {
+		// List what will hold not used (not applied) TransportSocketMatches to use with direct merging later
+		newPatchTsms := make([]*cluster.Cluster_TransportSocketMatch, 0, len(c.GetTransportSocketMatches()))
+
+		for _, cpTsm := range cpValueCast.GetTransportSocketMatches() {
+			if cpTsm.GetName() == "" || (cpTsm.GetTransportSocket() == nil && cpTsm.GetMatch() == nil) {
+				// Don't try merging if TransportSocketMatches' TransportSocket and Match are both null or
+				// name is empty - name is used to match TransportSockets in patch to the one in live Cluster and
+				// if TransportSocket and Match are both null there's nothing to patch
+				newPatchTsms = append(newPatchTsms, cpTsm)
+				continue
+			}
+
+			tsmMerged := false
+
+			for _, tsm := range c.GetTransportSocketMatches() {
+				if tsm.GetName() != cpTsm.GetName() {
+					// Skip unmatched names
+					continue
+				}
+
+				if cpTsm.GetMatch() != nil {
+					// Replace Match if defined in patch
+					tsm.Match = cpTsm.Match
+					tsmMerged = true
+				}
+
+				// Merge TransportSockets
+				if tsm.GetTransportSocket() != nil && cpTsm.GetTransportSocket() != nil {
+					if cpTsm.GetTransportSocket().Name == tsm.GetTransportSocket().Name {
+						// Merge when it's the same type
+						err := patchTransportSocket(tsm.GetTransportSocket(), cpTsm.GetTransportSocket())
+						if err != nil {
+							return err
+						}
+					} else {
+						// Replace when type mismatch
+						tsm.TransportSocket = cpTsm.TransportSocket
+					}
+					tsmMerged = true
+				}
+			}
+
+			if !tsmMerged {
+				newPatchTsms = append(newPatchTsms, cpTsm)
+			}
+		}
+
+		if len(newPatchTsms) != len(cpValueCast.TransportSocketMatches) {
+			cpValueCast.TransportSocket = nil
+		}
+		cpValueCast.TransportSocketMatches = newPatchTsms
 	}
 
 	// Check if cluster patch has a transport socket.
 	if cpValueCast.GetTransportSocket() == nil {
-		return false, nil
+		return nil
 	}
 	var tsmPatch *core.TransportSocket
 
-	// First check if the transport socket matches with any cluster transport socket matches.
+	// Check if the transport socket matches with any cluster transport socket matches.
```

Comment:
not related to this patch, just to mark the below handling is very hacky.  I am even confused after reading carefully